pauliopt/phase/cx_circuits.py

```diff
@@ -4,13 +4,115 @@
+        List[int]: The output permutation of the qubit mapping. Equals [0..n] if reallocate==False.
+    """
+    try:
+        # pylint: disable = import-outside-toplevel
+        import networkx as nx
+    except ModuleNotFoundError as _:
+        raise ModuleNotFoundError("You must install the 'networkx' library.")
+    try:
+        # pylint: disable = import-outside-toplevel
+        import galois
+    except ModuleNotFoundError as _:
+        raise ModuleNotFoundError("You must install the 'galois' library.")
+    cnots = []
+    m = np.asarray(matrix, dtype=np.int32)
+    def add_cnot(ctrl, trgt, m, cnots):
+        m[ctrl,:] ^= m[trgt,:]
+        if parities_as_columns:
+            cnots.append((ctrl,trgt))
+        else:
+            cnots.append((trgt, ctrl))
+    def choose_row(options, m):
+        return options[np.argmin([sum(m[o]) for o in options])]
+    def choose_column(options, row, m):
+        if reallocate:
+            return options[np.argmin([sum(m[:,o]) if m[row, o]==1 else topology.num_qubits for o in options])]
+        return row
+    
+    qubits_to_process = list(range(topology.num_qubits))
+    columns_to_eliminate = list(range(topology.num_qubits))
+    new_mapping = [-1]*topology.num_qubits
+    while len(qubits_to_process) > 1:
+        # Pick the pivot location
+        possible_qubits = topology.non_cutting_qubits(qubits_to_process)
+        # TODO does this change when parities_as_rows?
+        row = choose_row(possible_qubits, m)
+        col = choose_column(columns_to_eliminate, row, m)
+
+        # Reduce the column
+        col_nodes = [i for i in qubits_to_process if m[i, col]==1]
+        if m[row, col] == 0:
+            col_nodes.append(row)
+            
+        steiner_tree = topology.steiner_tree(col_nodes, qubits_to_process)
+        if len(col_nodes) > 1:
+            traversal = list(reversed(list(nx.bfs_edges(steiner_tree, source=row))))
+            for parent, child in traversal:
+                if m[parent, col] == 0:
+                    add_cnot(parent, child, m, cnots)
+            # TODO possibly pick the row here now that the full steiner_tree has a 1
+            for parent, child in traversal:
+                add_cnot(child, parent, m, cnots)
+        assert(sum(m[:,col]) == 1 )
+        assert(m[row, col] == 1)
+
+        # Reduce the row
+        ones_in_the_row = [i for i in columns_to_eliminate if m[row, i]== 1]
+        if len(ones_in_the_row) > 1:
+            submatrix = [[ m[i,j] for i in qubits_to_process if i != row] for j in columns_to_eliminate if j != col]
+            A = galois.GF(2)(submatrix)
+            A_inv = np.linalg.inv(A)
+            b = galois.GF(2)([m[row, i] for i in columns_to_eliminate if i != col], dtype=np.int32)
+            X1 = np.matmul(A_inv, b)
+            # Add the row that we removed back in for easier indexing.
+            X = np.insert(X1, qubits_to_process.index(row), 1)
```

Comment:
Minor thing, but does this work:
```python
remaining_qubits_to_process = list(filter(lambda x: x != row, qubits_to_process))
remaining_columns_to_eliminate = list(filter(lambda x: x != row, qubits_to_process))
A = galois.GF2(m[remaining_qubits_to_process, remaining_columns_to_eliminate])
B = galois.GF2(m[: , remaining_columns_to_eliminate], dtype=np.int32)
X = np.linalg.inv(A) @ B
# Add the row that we removed back in for easier indexing.
X = np.insert(X1, qubits_to_process.index(row), 1)
```
It reflects your algorithm much better!

Reply:
Ahh sorry my bad, this way will not work, but:
```python
A = np.delete(A, row, axis=0)
```
should do the trick

Reply:
Ah, yes, something like that could work. 
Although in this case, you'd need to be very careful that these calls don't accidentally change the original matrix `m`. That's why I opted for explicit copying of the data. I did not feel like debugging shenanigans.

Also, the pseudocode in the paper was adapted from the code I wrote here ;)